### DISCUSSION TITLE
HTTP Handlers fail on non-string request bodies

### DISCUSSION THREAD

**Discussion by StachuDotNet** [darklang/dark#4325] [Feedback]
Currently, Dark HTTP handlers automatically convert request bodies into strings, even if it's not valid Unicode. This means users can't send bytes.

An immediate use case for a user sending bytes is to support [static asset uploads in Dark](#4259), replacing the old/current OCaml-based upload. A good first goal here is to simply unblock this. (getting to 'good enough for internal use').

This discussion is largely happening in in a [google doc](https://docs.google.com/document/d/1vO0O8F24YGcF5DVixHdmZM-c66L_2vhrzJAOTgr2quY/edit?usp=sharing). While Paul and I are largely communicating there, external feedback is welcome here.

**Comment by StachuDotNet** [comment]
This is related to a larger problem of users not being able to control their HTTP handlers' pipelines.

The processing of a request to a user's HTTP handler is largely affected by logic within BwdServer and the HTTP Middleware. This includes the deserialization of request data, interpretation of content types, the serialization of responses, etc. While the goal of this project is not to solve for this larger problem, any wins towards this would be worthwhile.

**Comment by StachuDotNet** [comment]
Is this an opportunity / appropriate time to add DByte as a type, and generally switch from DBytes to List<DByte> throughout our code?

**Reply by StachuDotNet** [reply]
At least, before exposing the new middleware to users? I can't think of why a user would specifically want to work in a single Byte, but it feels appropriate to grant that control upon release rather than later

**Reply by pbiggar** [reply]
We won't be able to switch now because the memory usage from wrapping each byte in an object (for DByte) would be cost prohibitive. This will also be a relatively easy migration later.
